### DISCUSSION TITLE
Does Prisma have a built-in generated type for picking only scalar fields from `[Model]CreateInput`?

### DISCUSSION THREAD

**Discussion by vietan0** [prisma/prisma#24001] [Q&A] [answered]
### Question

When user is creating a `Project`, I want my user input (e.g. `req.body`) to have type-safety when writing middlewares, and maybe have a `zod` schema to validate it.

For example I have a model `Project`:
```prisma
model Project {
  id        String   @id @default(uuid()) @db.Uuid
  name      String   @db.VarChar(255)
  body      String
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  userId    String   @db.Uuid
  user      User     @relation(fields: [userId], references: [id])
  tasks     Task[]
}
```

The generated type `ProjectCreateInput` is:
```ts
type Prisma.ProjectCreateInput = {
    id?: string | undefined;
    name: string;
	body: string;
    createdAt?: string | Date | undefined;
    updatedAt?: string | Date | undefined;
    user: Prisma.UserCreateNestedOneWithoutProjectsInput;
    tasks?: Prisma.TaskCreateNestedManyWithoutProjectInput | undefined;
}
```

I'll use `connect` and pass `userId` from somewhere else (cookie, header,...). 
So user input (`req.body`) only needs
```ts
{
    id?: string | undefined;
    name: string;
	body: string;
    createdAt?: string | Date | undefined;
    updatedAt?: string | Date | undefined;
}
```

My problem is I can't use `ProjectCreateInput` to validate `req.body`, since it includes relation fields (`user`, `tasks`). In this one case I can use `Exact` or `Pick<ProjectCreateInput, { name: string, body: string, ... }>`, but I won't be able to if I have 15 more fields to pick.

So I need a type to pick out "scalar fields only". Is there a built-in type?

**Comment by ludralph** [comment, marked as answer]
Hi @vietan0 👋 

Prisma does not provide a built-in type that includes only the non-relation scalar fields. However, you can create such a type yourself using TypeScript's utility types. For example, you can use the `Omit` utility type to remove the relation fields from the `ProjectUncheckedCreateInput` type:
```ts
 export type ProjectUncheckedCreateInput = {
    id?: string
    name: string
    body: string
    createdAt?: Date | string
    updatedAt?: Date | string
    userId: string
    tasks?: TaskUncheckedCreateNestedManyWithoutProjectInput
  }
```

```ts
type ProjectScalarInput = Omit<ProjectUncheckedCreateInput, 'userId' | 'tasks'>;
```
This will give you a type that includes only the non-relation scalar fields. You can use this type to validate your `req.body` input.
